# Prueba 1 - Alejandro Florido Reyes

*Este documento contiene mis respuestas a los problemas planteados en la Prueba 1 de la asignatura Digital Transformation.*

*Para facilitar la lectura he colocado las respuestas cortas y explicaciones pedidas en cuadros amarillos y mis reflexiones adicionales en cuadros verdes. Además, he colocado notas en negrita a lo largo del texto cuando encontraba algo que me resultara ambiguo o extraño, explicando cómo lo he solucionado.*

## Problema 1.
Ejecuta el siguiente código desde el editor y explica qué produce:

In [10]:
type(5)
print(3.0-1)

2.0


Pero este resultado no está entre las respuestas, por lo que intuyo que hay que escribir ambas líneas por separado, de manera que se muestren ambos resultados (al haber varias operaciones en la ventana, todas aquellas que no sean la última y no estén en un print no se imprimen, por lo que type(5) no se imprime).

In [4]:
type(5)

int

El comando type() te indica el tipo del dato que tiene en el paréntesis. En este caso te indica que 5 es un entero (integer).

In [5]:
print(3.0-1)

2.0


<div class="alert alert-block alert-warning">
<b>Respuesta simplificada</b> 
    
3. int y 2.0. 

</div>

**Nota. Había que ejecutar ambas líneas por separado para que la respuesta se encontrara entre las propuestas en el documento.**

<div class="alert alert-block alert-success">
<b>Respuesta ampliada </b> 

El comando print() imprime la operación dentro del paréntesis. Dicha operación es entre un float y un integer y, como resultado de la operación, Python devuelve float en lugar de un integer. 

Esto último se puede entender si pensamos en una operación más general, porque en esta el resultado podía ser un entero, pero si la operación hubiera sido "3.1-1" el resultado no tendría más remedio que ser un float así que, para generalizar y mantener un criterio, Python hace que el resultado de una operación entre un float y un integer sea siempre un float.
</div>

## Problema 2.

Python vs. Matemáticas

¿Qué está permitido en Python?

<div class="alert alert-block alert-warning">
<b>Respuesta simplificada</b> 
    
4. xy = 2. 

</div>

<div class="alert alert-block alert-success">
<b>Respuesta ampliada </b> 

In [17]:
x + y = 2

SyntaxError: cannot assign to operator (<ipython-input-17-130ae22e9890>, line 1)

<div class="alert alert-block alert-success">
Devuelve error porque x e y no se han definido previamente.

In [18]:
x*x = 2

SyntaxError: cannot assign to operator (<ipython-input-18-561bdc127256>, line 1)

<div class="alert alert-block alert-success">
Devuelve error porque x no se ha definido previamente.

In [19]:
2 = x

SyntaxError: cannot assign to literal (<ipython-input-19-10ac2aee613c>, line 1)

<div class="alert alert-block alert-success">
Devuelve error porque, por un lado, no se puede usar un número para definir una variable (y las variables a definir van a la izquierda) y, por otro lado, porque x no está definida. 

La misma línea escrita justo al revés no devolvería error.

In [15]:
xy = 2

<div class="alert alert-block alert-success">
No devuelve error, ya que xy es una cadena de caracteres tan válida como cualquier otra para definir una variable. 

## Problema 3. 

¿Qué aparece como resultado de las siguientes líneas de código?

In [20]:
usa_oro = 46
uk_oro = 27
rumania_oro = 1

total_oro = usa_oro + uk_oro + rumania_oro
print(total_oro)

rumania_oro += 1
print(total_oro)

74
74


<div class="alert alert-block alert-warning">
<b>Respuesta simplificada</b>
    
1. 74 y 74.

</div>

<div class="alert alert-block alert-success">
<b>Respuesta ampliada </b> 

Como no se ha actualizado la variable total_oro después de aumentar el valor de rumania_oro, el valor que imprime el segundo print es el mismo que imprime el primero. El código devuelve 2 valores porque hay 2 prints.
</div>

## Problema 4.

1. ¿Qué va a ocurrir tras llamar a "u"?

In [31]:
comienzo = "par"
repetir = "aguas"
u = comienzo + (repetir + " ")*4

<div class="alert alert-block alert-warning">
<b>Respuesta simplificada</b>
    
1. paraguas aguas aguas aguas.

</div>

**Nota. Existe una pequeña posibilidad de que el autor del código no haya escrito " " en el código, sino "", en cuyo caso la respuesta sería 2. paraguasaguasaguasaguas,** pero he descartado esa posibilidad porque:
    1. En ese caso no habría tenido por qué escribir explícitamente ' + "" ' dentro del paréntesis.
    2. Por lo que se ve en el pdf, ciertamente parece haber un espacio entre ambas comillas.

<div class="alert alert-block alert-success">
<b>Respuesta ampliada </b> 

Al haber un paréntesis que incluye la operación repetir + " ", esto es lo primero que se une. A continuación, se copia 4 veces, porque la multiplicación tiene prioridad sobre la suma y, finalmente, la suma hace que comienzo ("par") se una a la primera de las 4 "aguas " generadas previamente. La siguiente línea de código muestra que este es, en efecto, el resultado:
</div>

In [32]:
u

'paraguas aguas aguas aguas '

2. ¿Qué pasa en cada paso?

In [35]:
hi = "hola"
name = "ana"
saludo= hi + name  
print(saludo)
#saludo = hi + " " + name
#print(saludo)
#mucho = hi + (" " + name)*3
#print(mucho)


holaana


<div class="alert alert-block alert-warning">
<b>Explicación</b>
    
* La primera línea guarda en una variable el string "hola". 
    
* La segunda guarda en otra variable el string "ana". 
    
* La tercera guarda en una nueva variable ambos strings unidos. 
    
* La cuarta imprime la nueva variable, esto es, ambos strings juntos, resultando el string "holaana", ya que no se ha colocado un espacio entre ambos strings en el proceso de unión. 

* Las líneas de la quinta a la octava son comentarios, por lo que las operaciones que contienen no son tenidas en cuenta a la hora de ejecutar el código. Entre las operaciones de estos comentarios encontramos la variable "saludo" modificada para que sí imprimiera "hola ana" (con el espacio), otra variable que devolvería "hola ana ana ana" y dos prints que, si no fuera porque las cuatro líneas están comentadas, haría que se imprimieran dichos mensajes.

</div>

## Problema 5.

¿Qué sucede tras ejecutar los comandos siguientes?

In [1]:
tiempo_trabajo = 15
tiempo_descanso = 8
print(tiempo_descanso > tiempo_trabajo)
conducir = True
beber = False
ambos = conducir and beber
print(ambos)

False
False


<div class="alert alert-block alert-warning">
<b>Respuesta simplificada</b>
    
* False y False

</div>

<div class="alert alert-block alert-success">
<b>Respuesta ampliada </b> 

El primer False viene de que lo que se imprime en el primer print es EL RESULTADO de la operación lógica (tiempo_descanso > tiempo_trabajo) que se realiza antes de llevar a cabo el print y que es False porque 8 NO es mayor que 15.

El segundo False se debe a que, en el segundo print, se imprime el resultado de la operación lógica "conducir and beber" (almacenado en la variable "ambos"). Como conducir es True y beber es False, "conducir and beber" es equivalente a "True and False", pero el resultado de esta operación booleana es False (ya que para que el resultado fuera True con un "and", sería necesario que tanto "beber" como "conducir" fueran True), de modo que el segundo print también devuelve False.
</div>

## Problema 6

¿Qué pasa si (en el siguiente código) x=0 e y=5?

In [37]:
x = float(input("Introduce un número para x: "))
y = float(input("Introduce un número para y: "))

if x == y:
    if y != 0:
        print("x/y es", x/y)
elif x < y:
    print("x es pequeño")
else:
    print("y es pequeño")
                

Introduce un número para x: 0
Introduce un número para y: 5
x es pequeño


<div class="alert alert-block alert-warning">
<b>Respuesta simplificada</b>
    
1. x es pequeño.

</div>

<div class="alert alert-block alert-success">
<b>Respuesta ampliada </b> 

El primer False viene de que lo que se imprime en el primer print es EL RESULTADO de la operación lógica (tiempo_descanso > tiempo_trabajo) que se realiza antes de llevar a cabo el print y que es False porque 8 NO es mayor que 15.

El segundo False se debe a que, en el segundo print, se imprime el resultado de la operación lógica "conducir and beber" (almacenado en la variable "ambos"). Como conducir es True y beber es False, "conducir and beber" es equivalente a "True and False", pero el resultado de esta operación booleana es False (ya que para que el resultado fuera True con un "and", sería necesario que tanto "beber" como "conducir" fueran True), de modo que el segundo print también devuelve False.
</div>

## Problema 7

Completar el código. Código de partida:

pi = 3.14159

radio = 2.2

#escribir la ecuación del área del círculo y que se vea el resultado.

In [8]:
pi = 3.14159
radio = 2.2
# Escribir la ecuación del área del círculo y que se vea el resultado.
# La ecuación del área del círculo es pi*radio**2, por lo tanto:

print(f'La ecuación del área del círculo es pi*radio**2 y, si sustituimos los valores del código en la ecuación obtenemos: {pi*radio**2}')

La ecuación del área del círculo es pi*radio**2 y, si sustituimos los valores del código en la ecuación obtenemos: 15.205295600000001


Otra opción *para que no se vea el resultado "cortado"* es hacer lo siguiente:

In [10]:
pi = 3.14159
radio = 2.2
# Escribir la ecuación del área del círculo y que se vea el resultado.
# La ecuación del área del círculo es pi*radio**2, por lo tanto:

print('La ecuación del área del círculo es pi*radio**2 y, si sustituimos los valores del código en la ecuación obtenemos: ')

print(pi*radio**2)

La ecuación del área del círculo es pi*radio**2 y, si sustituimos los valores del código en la ecuación obtenemos: 
15.205295600000001


## Problema 8

¿Qué pasa en cada print? ¿Qué aporta cada solución?

In [40]:
x = 1
print(x)
x_str= str(x)
print("mi número favorito es", x, ".", "x=", x)
print(" mi número favorito es ", x_str+ "." + "x=" + x_str)
print(" mi número favorito es " + x_str+ "." + "x=" + x_str)

1
mi número favorito es 1 . x= 1
 mi número favorito es  1.x=1
 mi número favorito es 1.x=1


<div class="alert alert-block alert-warning">
<b>Explicación</b>
    
Partamos desde el principio del código para que todo esté claro en la explicación de los prints. La primera línea almacena el valor 1 (integer) en la variable x, la segunda imprime dicho valor y la tercera cambia el tipo de x (es decir, del integer 1) a string y almacena ese string en la variable x_str.

Dicho eso, ¿qué ocurre en cada print? 

* Primer print: como añadimos 1 en forma de integer (x), vamos separando las partes strings de las partes donde metemos el 1 mediante comas, de forma que las diferentes partes separadas en el código se unen cuando ejecutamos el print. Obsérvese que, por defecto, se añade un espacio a los lados del 1 donde había una coma (podemos decir que, al ejecutar el comando, la coma se transforma en un espacio que aparece entre las diferentes partes que unía dicha coma). El resultado es: "mi número favorito es 1 . x= 1".

* Segundo print: en este caso añadimos el 1 en forma de string. La primera vez que aparece el uno lo añadimos como en el print anterior, separando ambas partes del texto con una coma y eso hace que, además del espacio que existe después del "es" veamos un segundo espacio antes del primer 1 de la frase. El resto de las partes de la frase, sin embargo, se unen mediante "+", por lo que el resto de partes se unen tal cual (el "+" no se sustituye por un espacio al ejecutar el print, simplemente une las dos partes). El resultado es: " mi número favorito es  1.x=1".

* Tercer print: en este caso se usa el 1 en forma de string y todas las uniones son mediante "+", de modo que no se añaden espacios adicionales a los strings unidos al ejecutar el comando print. El resultado es: " mi número favorito es 1.x=1".

¿Qué aporta cada solución? 

* El primer print nos enseña que podemos unir strings con integers usando comas para separar ambos tipos de dato en un comando print(). Además, muestra que una coma se transforma en un espacio al ejecutar el print, por lo que tenemos que tenerlo en cuenta si queremos que la frase a imprimir no tenga más espacios de la cuenta. 

* El segundo print muestra que los strings se pueden unir entre sí tanto con las comas como con signos de suma (+) y que estos últimos no se transforman en un espacio adicional como ocurría con las comas, sino que simplemente unen los strings que tienen a su izquierda y su derecha.

* El tercer print no aporta nada nuevo. Si acaso que en el segundo print no había necesidad de usar la coma, se podía hacer todo con el "+".

</div>

## Problema 9

¿Qué ocurre?

In [1]:
text = '5'
print(5*text)
num = 5
print(5*num)


55555
25


<div class="alert alert-block alert-warning">
<b>Respuesta simplificada </b>
    
3. 55555 y 25

</div>

<div class="alert alert-block alert-success">
<b>Respuesta ampliada </b> 

La variable text almacena el string 5, por lo tanto, 5*text escribe 5 veces seguidas el caracter 5 (esto es, 55555) por lo que eso es lo que imprime el primer print. 

Por otra parte, num es una variable que almacena el integer 5 (un número), por lo tanto, en este caso 5*num es 5x5 = 25 (multiplica el integer almacenado por 5) y eso es lo que imprime el segundo print.

Por lo tanto la respuesta es la 3.
</div>

## Problema 10

¿Qué ocurre si se introduce 12314588134 y una a?

In [3]:
num = int(input("Un número, por favor: "))
if num % 2 == 0:
    print("par")
else:
    print("impar")

Un número, por favor: 12314588134
par


In [6]:
num = int(input("Un número, por favor: "))
if num % 2 == 0:
    print("par")
else:
    print("impar")

Un número, por favor: a


ValueError: invalid literal for int() with base 10: 'a'

In [5]:
num = int(input("Un número, por favor: "))
if num % 2 == 0:
    print("par")
else:
    print("impar")

Un número, por favor: 'a'


ValueError: invalid literal for int() with base 10: "'a'"

<div class="alert alert-block alert-warning">
<b>Respuesta simplificada</b>
    
4. Par y error

</div>

<div class="alert alert-block alert-success">
<b>Respuesta ampliada </b> 

Al introducir 12314588134 como input, este se almacena en la variable num. En consecuencia, num es un número par y, por lo tanto, cumple la condición del if (el resto que se obtiene al dividir 12314588134 entre 2 es 0) y se imprime "par". Como se cumple la condición del if, la operación del else no se lleva a cabo.

Al introducir una a (ya sea en forma de string, 'a', o directamente) se obtiene un error tipo ValueError, ya que el valor que hemos almacenado en la variable num (a o 'a') no es numérico y, por lo tanto, la operación num % 2 no se puede llevar a cabo y no se puede evaluar el if. Para poder evaluar la condición del if, num debe ser un número (integer, float, etc.).
</div>